### PR TITLE
#182631416 - USERS SHOULD BE ABLE TO BOOKMARK ARTICLES FOR READING LATER

### PR DESCRIPTION
**Pull request by dalienst**
#### What does this Pull Request do?

- Enable user to bookmark articles

#### Description of Task to be completed?

- Have the following endpoints working

1. POST: api/articles/bookmarks/

2. GET: api/articles/bookmarks/

#### How should this be manually tested?

- Using postman, test every endpoint above

#### Any background context you want to provide?

- The ability of users to bookmark articles and read them later.

#### What are the relevant pivotal tracker?

- [182631416](https://www.pivotaltracker.com/n/projects/2570600/stories/182631416)

#### Screenshots (if appropriate)

- 
Screenshot from 2022-08-03 23-48-56

#### Questions:

- N/A

### CODE REVIEW

File: app/articles/models.py [hunk 1]
```diff
@@ -54,3 +54,16 @@ def __str__(self) -> str:
 def slug_pre_save(sender: Any, instance: Any, **kwargs: Any) -> None:
     if instance.slug is None or instance.slug == "":
         instance.slug = slugify(f"{instance.title}-{instance.post_id}")
+
+
+class ArticleBookmark(models.Model):
```
Comment (mainamuriuki274) [inline app/articles/models.py:59]: use the timestamped model

File: app/articles/serializers.py [hunk 2]
```diff
@@ -81,3 +81,25 @@ def create(self, validated_data: Any) -> Any:
...
+    class Meta:
+        model = ArticleBookmark
+        fields = (
+            "article",
+            "user",
+        )
+
+    def create(self, validated_data: Any) -> Any:
+        request = self.context["request"]
+        instance = ArticleBookmark.objects.get_or_create(
```
Comment (mainamuriuki274) [inline app/articles/serializers.py:102]: As a BTW, another way you could have done this is using a hidden field with currentUserDefault

File: app/articles/tests/test_views.py [hunk 3]
```diff
@@ -108,3 +108,48 @@ def test_delete_article_without_authentication(self) -> None:
...
+            body=fake.paragraph(),
+        )
+
+        response = self.client.post(
+            reverse("bookmark"),
+            data={"article": data.post_id},
+            **self.bearer_token,
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:148]: assert more than once, eg that the data returned matches the article/user

File: app/articles/tests/test_views.py [hunk 4]
```diff
@@ -108,3 +108,48 @@ def test_delete_article_without_authentication(self) -> None:
...
+
+    def test_get_bookmark(self) -> None:
+        response = self.client.get(
+            reverse("bookmark"),
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:155]: assert more than once eg the one gotten is the one you want

File: app/articles/tests/test_views.py [hunk 5]
```diff
@@ -108,3 +108,48 @@ def test_delete_article_without_authentication(self) -> None:
...
+
+        response = self.client.post(
+            reverse("bookmark"),
+            data={"article": data.post_id},
+            **self.bearer_token,
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_get_bookmark(self) -> None:
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:150]: add a docsting here

File: app/articles/tests/test_views.py [hunk 6]
```diff
@@ -108,3 +108,48 @@ def test_delete_article_without_authentication(self) -> None:
...
+    def bearer_token(self) -> dict:
+        login_url = reverse("login")
+        response = self.client.post(
+            login_url,
+            data={"email": self.user.email, "password": self.password},
+        )
+        token = json.loads(response.content).get("access")
+        return {"HTTP_AUTHORIZATION": f"Bearer {token}"}
+
+    def test_bookmark_article(self) -> None:
```
Comment (mainamuriuki274) [inline app/articles/tests/test_views.py:135]: add a docstring

File: app/articles/urls.py [hunk 7]
```diff
@@ -9,4 +13,6 @@
         ArticleDetailView.as_view(),
         name="article-detail",
     ),
+    path("bookmarks/", ArticleBookmarkView.as_view(), name="bookmark"),
+    # path("bookmarks/delete/<str:article>/", BookmarkDestroy.as_view(), name="bookmark"),
```
Comment (mainamuriuki274) [inline app/articles/urls.py:17]: remove unused comments

File: app/articles/views.py [hunk 8]
```diff
@@ -38,3 +40,15 @@ def delete(self, request: Request, *args: Any, **kwargs: Any) -> Response:
...
+class ArticleBookmarkView(generics.ListCreateAPIView):
+    serializer_class = ArticleBookmarkSerializer
+    queryset = ArticleBookmark.objects.all()
+    permission_classes = [
+        IsAuthenticated,
+    ]
+
+    def get_queryset(self) -> Any:
+        queryset = super().get_queryset()
+        return queryset.filter(user=self.request.user)
```
Comment (mainamuriuki274) [inline app/articles/views.py:54]: Nice

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

**Review by ChrisOnsando — COMMENTED** [review]
Changes made, articles bookmarking working fine

**Review by mainamuriuki274 — APPROVED** [review]

**Pull request merged**